File: HW6/GTUSet.cpp
#include <iostream>
#include "GTUSet.h"
#include <memory>
#include <stdexcept>

using namespace std;

namespace GTUSetAndMap{

	template<class T>
	GTUSet<T>::GTUSet()
			: GTUSetBase<T>()
	{	
		container = shared_ptr<T>(new T[this->get_capacity()],std::default_delete<T[]>());
	}		

	template<class T>
	GTUSet<T>::GTUSet( int container_size )
			: GTUSetBase<T>( container_size )
	{
		container = shared_ptr<T>(new T[ this->get_capacity()],std::default_delete<T[]>());
	}

	template<class T>
	T& GTUSet<T>::operator[](int index){

		if ( index >= this->get_used() ){
			cout << "illegal index value" << endl;
			exit(1);
		}
		else
			return container.get()[index];
	}

	template<class T>		
	bool GTUSet<T>::empty(){

		if( this->get_used() == 0 )
			return true;
		else
			return false;
	}


	template<class T>
	void GTUSet<T>::insert( T theValue ){
		
		int found = 0;

		for (int i = 0; i < this->get_used(); ++i)
		{
			if( theValue == container.get()[i] )
				found = 1;

		}
		
		if( !found ){
			container.get()[this->get_used() ] = theValue;
			this->set_used(1);
		}
		else
			throw std::invalid_argument("The element is already in the set.\n");
			

		this->sort();
	}

	template<class T>
	void GTUSet<T>::sort(){

		for ( int i = 0; i < this->get_used() - 1 ; ++i )
		{	
			for (int j = i+1; j < this->get_used(); ++j)
			{	
				if ( container.get()[j] < container.get()[i]  ){
					T temp = container.get()[i];
					container.get()[i] = container.get()[j];
					container.get()[j] = temp;
				}									
			}
		}
	}

	template<class T>
	void GTUSet<T>::erase(T theValue){
		int size;
		T *arr;
		for ( int i = 0; i < this->get_used() ; ++i ){

			if ( container.get()[i] == theValue )
			{	
				size = this->get_used() - (i+1);
				arr = new T[size];

				for (int j = i+1, k = 0; j < this->get_used(); ++j, ++k)
				{
					arr[k] = container.get()[j];
				}

				for (int j = 0, k = i; j < size; ++j, ++k)				
					container.get()[k] = arr[j];

				this->set_used(2);
			}

		}
	}

	template<class T>
	void GTUSet<T>::print(){

		for (int i = 0; i < this->get_used(); ++i)
			cout << container.get()[i] << "  ";

		cout << endl;

	}

	template<class T>
	T* GTUSet<T>::begin(){
		T* iter;
		iter = &(*container);
		return iter;
	}		


	template<class T>
	T* GTUSet<T>::end(){
		T* iter;
		iter = &(container.get()[this->get_used()-1]);
		return iter;
	}

	template<class T>
	int GTUSet<T>::size(){	return this->get_used();	}

	template<class T>
	int GTUSet<T>::max_size(){	return this->get_capacity();	}

	template<class T>
	int GTUSet<T>::count(T theValue){

		int count = 0;

		for (int i = 0; i < this->get_used(); ++i)
		{
			if( theValue == container.get()[i] )
				++count;
		}

		return count;
	}

	template<class T>
	T* GTUSet<T>::find(T theValue){

		T* iter;
		int found = 0;

		for (int i = 0; i < this->get_used(); ++i)
		{
			if( theValue == container.get()[i] ){
				iter = &(container.get()[i]);
				found = 1;
			}				
		}

		if (found == 1)
			return iter;
	}
	
	template<class T>
	void GTUSet<T>::clear(){

		this->set_used(0);

	}


}

File: HW7/GTUSetInt_Test/src/gtusetınt_test/GTUSet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
        
package gtusetınt_test;

import java.lang.reflect.Array;
import java.util.Arrays;
/**
 *  GTUSet class i GTUSetInt interface inden implement edilir
 * @author Burcu SUVAK OZTURK 141044079
 * 
 */

public class GTUSet<T> implements GTUSetInt<T>{
    
    public class GTUIterator<T>{
               
       private int currentIndex = 0;
       private int currentSize;
       
       public GTUIterator(){           
           currentSize = capacity;           
       }
       
        /**
        * @return currentIndex kapasiteden den kucukse true degilse false return eder
        */       
       public boolean hasNext(){
           
           return  currentIndex < currentSize && container[currentIndex] != null ;
       }
       
       /**
        * 
        * @return currentIndex 0 dan buyukse true degilse false return eder
        */       
       public boolean hasPrevious(){
            
           return currentIndex > 0;
       }
       
       /**
        * hasPrevious() fonksiyonu true dondururse bir oncesindeki eleman return edilir
        * @return iteratorun mevcut konumundan bir onceki elemani return eder
        */          
       public T previous(){
            if ( hasPrevious() )
                return (T)container[--currentIndex];
            else
                return null;           
       }
       
       /**
        *  hasNext() fonksiyonu true dondururse bir oncesindeki eleman return edilir
        * @return iteratorun mevcut konumundan sonraki elemani return eder
        */       
       public T next(){
            //if( "gtusetınt_test.GTUSet".equals(this.getClass().getName())  ) 
            if ( hasNext() )
                return (T)container[++currentIndex];
            else 
                return null;                                  
       } 
       
       /**
        * @return iteratorun konumlandigi index teki degeri dondurur
        */       
       public T getItem(){
           
           return (T)container[currentIndex];
       }
                
    }
    
    private int used;
    private int capacity;
    private final int increase;
    private Object[] container;

 /**
 * @param capacityValue Set in sahip oldugu maksimum eleman sayisi
 */
    public GTUSet(int capacityValue){
        capacity = capacityValue;
        used = 0;
        increase = 10;
        container = new Object[capacityValue];
    }
    /**
     * GTUSet class Constructor
     */
     public GTUSet(){
        capacity = 50;
        used = 0;
        increase = 10;        
        container = new Object[capacity];
    }
       
    /**
     *     
     * @return Set in kapasitesini döndürür
     */
    public int max_size(){      return capacity;    }
    
    /**
     * 
     * @return getUsed() Set içinde bulunan eleman sayısını döndürür
     */
    public int size(){  return getUsed();       }
    
    /**    
     * @return Set içinde bulunan eleman sayısını döndürür
     */
    
    public int getUsed(){
        return used;
    }
    /**
     *  Set in eleman sayısnı bir artırır.
     */
    public void setUsed(){
        ++used;       
        
    }
    
    /**
     * kapasite doldugunda Set in kapasite degerini arttrır.
     */
    public void setCapacity(){
        capacity += increase;
    }
    
    /**     
     * @return Set boş ise true degilse false döndürür
     */    
    @Override
    public boolean empty(){
        
        return getUsed() == 0;             
    }
    
    /**
     *  Set in tüm içeriğini temizler
     */
    @Override
    public void clear(){    
        container = null;
        used = 0;
    }
    
    /**
     *  insert() fonksiyonu set e eklanmek istenen eleman 
     *  eger set icinde mevcut degilse value degerini
     *  set e ekler ve used degerini bir artırır
     *  daha sonra elamanı uygun yere yerlestirebilmak için
     *  sort() fonksiyonunu cagırır ve kucukten buyuge
     *  sıralama isini yapar
     *  ancak set icinde bu eleman zaten 
     *  varsa bir invalid_value exception fırlatır
     * 
     * @param value set e eklamek istenilen deger
     */
    
    @Override
    public void insert(T value) throws IllegalArgumentException{

        int found = 0;
        
        if( getUsed() >= capacity ){
            setCapacity();
            System.out.println("\nTaking extra space...");
            Object[] backUpArray;
            
            backUpArray = new Object[getUsed()];
            for(int i = 0; i < getUsed(); ++i)
                backUpArray[i] = container[i];
            container = new Object[getUsed() + increase];
            
            for(int i = 0; i < getUsed(); ++i)
                container [i] = backUpArray[i];            
            
        }

        for( int i = 0; i < getUsed(); ++i ){
            if( container[i] == value )
                found = 1;  
        }                
               
        if( found == 0 ){
            container[getUsed()] = value;
            setUsed();
        }
        else
            throw new IllegalArgumentException("Invalid input...\n");
        
        sort();
    }
    
    /**
     *  printSet() set in tüm elemanlarını ekrana basar
     */
    public void printSet(){
        
        System.out.println(java.util.Arrays.toString( container ));
    }
   
    /**
    *   sort() fonksiyonu array eleman eklendikten sonra
    *   elemanları kücükten büyüğe dogru sıralar
    */    
    @Override
    public void sort(){
        
        Arrays.sort(container,0,getUsed());       
    }
    /**
     *  erase() fonksiyonu parametre olarak aldıgı value degeri 
     *  set in icinde varsa onu siler ve used degerini bir azaltır
     * @param value Setten silinmek istenen değer 
     */    
    @Override
    public void erase(Object value){
                
	for ( int i = 0; i < getUsed() ; ++i ){
            if ( container[i] == value )
            {	
                --used;
                
		for (int j = i; j < getUsed(); ++j)	                 
                    container[j] = container[j+1];                  	
				
            }
            container[getUsed()] = null;
	} 
   }
    
    /**
     * İntersection() fonksiyonu iki adet Set i karşılaştırır ve 
     * kesişimlerini döndürür.
     * @param other GTUSet objesi referansı
     * @return intersection objesini return eder
     */
    @Override
    public GTUSet<T> intersection(GTUSet<T> other){
        Object[] arr;
        int size;
        if ( this.getUsed() <= other.getUsed() )
            size = this.getUsed();
	else
            size = other.getUsed();
        arr = new Object[size];
        
        System.out.print("INTERSECTION : ");
        
        if ( this.getUsed() <= other.getUsed() ){
            for (int i = 0; i < this.getUsed(); ++i)
            {	
		for (int k = 0; k < other.getUsed(); ++k)
		{
                    if ( container[i] == other.container[k] )
                    {
			arr[i] = container[i];
      			System.out.printf("%d ",container[i]);

                    }
		}			                
            }
	}
        
        else{
            for (int i = 0; i < other.getUsed(); ++i)
            {	
		for (int k = 0; k < this.getUsed(); ++k)
		{
                    if ( other.container[i] == container[k] )
                    {
			arr[i] = other.container[i];
			System.out.printf("%d ",other.container[i]);
                    }
		}	
            }
	}        
        System.out.println();
        return other;
   }
    
    /**
     * @return container Set in referansını döndürür
     */   
    public Object[] returnItems(){
       
       return container;
    }
    
    /**
     * Bulunmak istenen degeri alır ve eger bulursa ıteratorunu dondurur
     * bulamazsa null return eder
     * @param theValue Set icinde bulunmak istenen deger
     * @return bulundugu yerin iteratoru
     */    
    @Override
    public GTUIterator<T> find(Object theValue){
        
        GTUIterator<T> iter = new GTUIterator();
        
        boolean found = false;

            for (int i = 0; i < getUsed() ; ++i)
            {   
		if( theValue == container[i] ){
                    found = true;
		}
                if(!found)

                    iter.next();

            }
                return iter;               
    }       
    
    public GTUIterator<T> begin(){
        
        GTUIterator<T> iter = new GTUIterator();

        return iter;
    }
       
    public GTUIterator<T> end(){
        
        GTUIterator<T> iter = new GTUIterator();
        for(int i = 0; i < getUsed()-1; ++ i)
            iter.next();
        return iter;
    }
}

